### PR TITLE
Blur mip

### LINKED ISSUE #724: Textures are blurring when in the distance in 0.6.0
### Description

0.6.0 added improvements to fix issues like the Moiré issue:  https://github.com/GoogleWebComponents/model-viewer/issues/632

It does a pretty good job:
Screen Shot 2019-08-27 at 1 39 45 PM

However on the left hand photo above, you notice some banding happening.

Looking at it closer, it looks as though the texture is blurring the further away from the camera it is. Seems like anisotropic filtering issue.
Screen Shot 2019-08-27 at 1 40 14 PM

Here's a gif of it happening (notice the detail in the stitching and zippers)
details

#### Live Demo
Compare both new and old viewers here: https://saber-soup.glitch.me/

### Browser Affected

- [ x ] Chrome 76.0.3809.100
- [ x ] Safari (desktop and mobile)  12.1.1 (14607.2.6.1.1)

### OS

- [x ] iOS (12.3)
- [x ] MacOS (10.14.5)

### Versions

- model-viewer: v0.6.0
- three.js: r105

### PR DESCRIPTION
**Pull request by elalish**
Fixes #724 

Well, I doubt this will completely fix #724, but we'll see. Mostly this is a refactor to improve and speed up PMREM. I condensed five shaders into one, which reduces our compile time significantly (a major contributor to first-environment startup time). I also reduced the number of draw calls by a factor of 3, which significantly reduced the time to run environment generation (a major contributor to next-environment startup time). This has the added benefit of anti-aliasing our skyboxes a bit, since I replaced three's EquirectangularToCubemap function.

### CODE REVIEW

**Comment by elalish** [comment]
@cdata @mrdoob I'm totally baffled here. Outside of puppeteer, all of the fidelity tests work great. Inside of puppeteer, my PMREM environments end up quite wrong, and it seems that the position attribute I'm passing into my RawShaderMaterial is getting stretched by a factor of 2. If I switch one line in my vertex shader to `gl_Position = vec4( 0.5 * position - 0.5, 1.0 );`, it ends up nearly correct (but only in puppeteer). Looking in the debugger everything appears to be as expected; the attributes all have the right values, the shaders are as expected. Does anyone have any ideas?

**Comment by elalish** [comment]
@cdata @mrdoob I finally figured it out! It was a DPR issue. When running offscreen renders I had neglected to set the pixel ratio back to one, hence why it only affected macbooks and puppeteer. It didn't occur to me that there was a scaling factor that applied to render target dimensions but not viewport dimensions.

**Comment by mrdoob** [comment]
> @cdata @mrdoob I finally figured it out! It was a DPR issue. When running offscreen renders I had neglected to set the pixel ratio back to one, hence why it only affected macbooks and puppeteer.

That makes A LOT of sense. Nice detective work!

File: src/model-viewer.ts [hunk 1]
```diff
@@ -23,6 +23,10 @@ import {StagingMixin} from './features/staging.js';
 import ModelViewerElementBase from './model-viewer-base.js';
 import {FocusVisiblePolyfillMixin} from './utilities/focus-visible.js';
 
+// Uncomment these lines to export PMREM textures in Glitch:
```
Comment (cdata) [inline src/model-viewer.ts]: 💯

File: src/three-components/PMREMGenerator.ts [hunk 2]
```diff
@@ -13,407 +13,489 @@
...
+        lod--;
+      }
+    }
+  }
 
-      const cubeUVRenderTarget =
-          new WebGLRenderTarget(3 * sizeMax, offsetY, params);
-      cubeUVRenderTarget.texture.name = 'PMREMCubeUVPacker.cubeUv';
-      cubeUVRenderTarget.texture.mapping = CubeUVReflectionMapping;
+  fromDefault(): WebGLRenderTarget {
```
[the remark this answers is not in the document]
Comment (elalish) [reply src/three-components/PMREMGenerator.ts]: Done.

File: src/three-components/PMREMGenerator.ts [hunk 3]
```diff
@@ -13,407 +13,489 @@
...
-        const {uniforms} = mipmapShader;
-        cubeCamera.renderTarget = cubeLods[i];
-        uniforms.outputEncoding.value = encodings[cubeLods[i].texture.encoding];
-        cubeCamera.update(renderer, mipmapScene);
-        uniforms.texelSize.value = 1.0 / cubeLods[i].width;
-        uniforms.envMap.value = cubeLods[i].texture;
-        uniforms.inputEncoding.value = encodings[cubeLods[i].texture.encoding];
-      }
-    };
+  fromScene(
```
[the remark this answers is not in the document]
Comment (elalish) [reply src/three-components/PMREMGenerator.ts]: Done.

File: src/three-components/PMREMGenerator.ts [hunk 4]
```diff
@@ -13,407 +13,489 @@
...
-}
-`;
+    this.renderer.setPixelRatio(dpr);
+    return cubeUVRenderTarget;
+  }
 
-class MipmapShader extends RawShaderMaterial {
-  constructor() {
-    super({
+  fromEquirectangular(equirectangular: Texture): WebGLRenderTarget {
```
[the remark this answers is not in the document]
Comment (elalish) [reply src/three-components/PMREMGenerator.ts]: Done.

File: src/three-components/PMREMGenerator.ts [hunk 5]
```diff
@@ -13,407 +13,489 @@
...
+const $pingPongRenderTarget = Symbol('pingP$pingPongRenderTarget');
+
+const $sceneToCubeUV = Symbol('sceneToCubeUV');
+const $equirectangularToCubeUV = Symbol('equirectangularToCubeUV');
+const $createRenderTarget = Symbol('createRenderTarget');
+const $applyPMREM = Symbol('applyPMREM');
+const $blur = Symbol('blur');
+const $halfBlur = Symbol('halfBlur');
+
+export class PMREMGenerator {
```
[the remark this answers is not in the document]
Comment (elalish) [reply src/three-components/PMREMGenerator.ts]: Done.

**Comment by mrdoob** [comment]
Wow! Looking incredible!

Before:
http://model-viewer-tester.glitch.me/0.6.0.html
Screen Shot 2019-10-08 at 9 00 52 AM

After:
http://model-viewer-tester.glitch.me/
Screen Shot 2019-10-08 at 9 01 08 AM

**Comment by mrdoob** [comment]
It's amazing how the handle of the boombox looks now 🤯

Before:
Screen Shot 2019-10-08 at 9 19 25 AM

After:
Screen Shot 2019-10-08 at 9 19 10 AM

**Comment by pushmatrix** [comment]
The mip checker boarding seems to still be happening: Screen Shot 2019-10-08 at 8 31 25 AM

Screen Shot 2019-10-08 at 8 36 04 AM

See here:
https://shelled-turret.glitch.me/

**Comment by pushmatrix** [comment]
@elalish I can confirm things do look significantly less wet! :D

**Pull request merged**